### PR TITLE
Reject chunks without timekey from candidates of destination of symlink

### LINKED ISSUE #1219: comparison of NilClass with 1473379200 failed
My stack:
* fluentd 0.14.6
* gem install json_pure fluent-plugin-loggly

My config:
* follow apache log files
* buffered, bulk upload to loggly

II get a bunch of errors like this:

```
fluentd_1  | 2016-09-09 07:37:34 +0000 [warn]: emit transaction failed: error_class=ArgumentError error="comparison of NilClass with 1473379200 failed" tag="apache.error"
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/out_file.rb:62:in `sort_by'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/out_file.rb:62:in `generate_chunk'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/buffer.rb:447:in `block in write_once'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/buffer.rb:447:in `write_once'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/buffer.rb:197:in `block in write'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/buffer.rb:196:in `each'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/buffer.rb:196:in `write'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/output.rb:665:in `block in handle_stream_with_custom_format'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/output.rb:610:in `write_guard'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/output.rb:664:in `handle_stream_with_custom_format'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/output.rb:602:in `execute_chunking'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/output.rb:544:in `emit_buffered'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/out_copy.rb:38:in `block in process'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/out_copy.rb:37:in `each'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/out_copy.rb:37:in `process'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/multi_output.rb:89:in `emit_sync'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/event_router.rb:153:in `emit_events'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/event_router.rb:90:in `emit_stream'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:310:in `receive_lines'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:420:in `wrap_receive_lines'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:579:in `on_notify'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:444:in `on_notify'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:425:in `attach'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:210:in `setup_watcher'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:231:in `block in start_watchers'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:218:in `each'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:218:in `start_watchers'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:204:in `refresh_watchers'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/plugin/in_tail.rb:156:in `start'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/root_agent.rb:138:in `block in start'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/root_agent.rb:127:in `block (2 levels) in lifecycle'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/root_agent.rb:126:in `each'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/root_agent.rb:126:in `block in lifecycle'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/root_agent.rb:113:in `each'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/root_agent.rb:113:in `lifecycle'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/root_agent.rb:137:in `start'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/engine.rb:211:in `start'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/engine.rb:175:in `run'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/supervisor.rb:677:in `run_engine'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/supervisor.rb:436:in `block in run_worker'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/supervisor.rb:606:in `main_process'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/supervisor.rb:431:in `run_worker'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/lib/fluent/command/fluentd.rb:271:in `<top (required)>'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.6/bin/fluentd:5:in `<top (required)>'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/bin/fluentd:23:in `load'
fluentd_1  |   2016-09-09 07:37:34 +0000 [warn]: /usr/bin/fluentd:23:in `<main>'
```

Test log record:
```
fluentd_1  | 2016-09-09 07:51:17 +0000 apache.access: {"test":1473407477293,"request_id":"V9Jp9awSAAYAAAAJftUAAACA","client":"10.0.2.89","host":"bmg","path":"/","query":"","method":"GET","status":"200","userAgent":"curl/7.50.1","duration":2,"request_size":72,"response_size":186}
```

I'm not sure if this is fluentd internal error or the output plugin.

### PR DESCRIPTION
**Pull request by tagomoris**
"symlink" feature is to link from symlink_path to the latest file chunk. Records with latest timekey will be appended into that file chunk.
On the other side, resumed file chunks might NOT have timekey, especially in the cases that resumed file chunks are generated by
Fluentd v0.12. These chunks will be enqueued immediately, and will be flushed soon.

So these files SHOULD NOT be a candidate of destination of symlink.

fixes #1219.

### CODE REVIEW

**Comment by tagomoris** [comment]
@repeatedly Could you review this change?

File: lib/fluent/plugin/out_file.rb [hunk 1]
```diff
@@ -59,7 +59,7 @@ def symlink_path=(path)
 
       def generate_chunk(metadata)
         chunk = super
-        latest_chunk = metadata_list.sort_by(&:timekey).last
+        latest_chunk = metadata_list.select{|m| m.timekey }.sort_by(&:timekey).last
```
Comment (repeatedly) [inline lib/fluent/plugin/out_file.rb]: Add comment is better for understanding why we pick up only chunks which have timekey data.

Comment (tagomoris) [inline lib/fluent/plugin/out_file.rb]: Added a comment.

**Comment by repeatedly** [comment]
LGTM

**Pull request merged**